### PR TITLE
test(robot): add uninstallation check test case

### PR DESCRIPTION
**Pull request by chriscchien**
#### Which issue(s) this PR fixes:

Issue #longhorn/longhorn#9222

#### What this PR does / why we need it:

Transfer manual test case `Uninstallation Checks` into robot framework

#### Special notes for your reviewer:

Having environment below variable set first if run on local
- `LONGHORN_INSTALL_METHOD` (`manifest` or `helm`) 
- `LONGHORN_REPO_BRANCH`
- `CUSTOM_LONGHORN_MANAGER_IMAGE` (if not using master-head)
- `CUSTOM_LONGHORN_ENGINE_IMAGE` (if not using master-head)
- `CUSTOM_LONGHORN_INSTANCE_MANAGER_IMAGE` (if not using master-head)
- `CUSTOM_LONGHORN_SHARE_MANAGER_IMAGE` (if not using master-head)
-`CUSTOM_LONGHORN_BACKING_IMAGE_MANAGER_IMAGE` (if not using master-head)

issue command
- `./run.sh -t "Uninstallation Checks" 

#### Additional documentation or context

[report.zip](https://github.com/user-attachments/files/16725623/report.zip)

### CODE REVIEW

**Comment by chriscchien** [comment]
Need to wait for clarification on https://github.com/longhorn/longhorn/issues/9303 before reviewing this PR. Currently, the PR only checks logs for `level=error` and `level=fatal` instead of looking for the word `error` alone.

```python
assert "level=error" not in logs
assert "level=fatal" not in logs
```

**Comment by chriscchien** [comment]
> Need to wait for clarification on [longhorn/longhorn#9303](https://github.com/longhorn/longhorn/issues/9303) before reviewing this PR. Currently, the PR only checks logs for `level=error` and `level=fatal` instead of looking for the word `error` alone.
> 
> ```python
> assert "level=error" not in logs
> assert "level=fatal" not in logs
> ```

This ticket is ready for review, I have changed the log check part to make sure no word `error` appeared in uninstall pod.

File: e2e/libs/uninstallation/uninstallation.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from node import Node
+from node_exec import NodeExec
+from k8s import k8s
+from utility.constant import LONGHORN_NAMESPACE
+from utility.constant import LONGHORN_UNINSTALL_JOB_LABEL
+from kubernetes import client
+
+import os
+
+class Uninstallation:
```
Comment (yangchiu) [inline e2e/libs/uninstallation/uninstallation.py:11]: The `Uninstallation` class is too specific. We can create a `Longhorn` abstract class includes `install`, `uninstall`, `upgrade` and other related interfaces. Then, classes like `LonghornKubectl`, `LonghornHelmChart`, `LonghornRancherChart` can extend the `Longhorn` class to implement these interfaces.

File: e2e/keywords/longhorn.resource [hunk 2]
```diff
@@ -43,3 +44,9 @@ Check Longhorn workload pods ${condition} annotated with ${key}
     Run Keyword IF    '${condition}' == 'not'    Should Not Be True    ${is_annotated}
     ...    ELSE IF    '${condition}' == 'is'    Should Be True    ${is_annotated}
     ...    ELSE    Fail    Invalid condition ${condition}
+
+Uninstall Longhorn ${LONGHORN_BRANCH} by ${INSTALL_METHOD}
```
Comment (yangchiu) [inline e2e/keywords/longhorn.resource:48]: We can use the `INSTALL_METHOD` environment variable to determine which class to instantiate, eliminating the need to explicitly specify `Uninstall Longhorn by XXX`. WDYT?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/tests/test_cases/uninstallation_checks.robot [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    ...    - Have one DR volume created (not activated) in another cluster with a volume in current cluster.
+    ...
+    ...    Test steps
+    ...    1. Uninstall Longhorn.
+    ...    2. Check the logs of the job longhorn-uninstall, make sure there is no error(skip this step if using helm).
+    ...    3. Check all the components of Longhorn from the namespace longhorn-system are uninstalled. E.g. Longhorn manager, Longhorn driver, Longhorn UI, instance manager, engine image, CSI driver etc.
+    ...    4. Check all the CRDs are removed kubectl get crds | grep longhorn.
+    ...    
+    ...    Not implemented Steps
+    ...    5. Check the backup stores, the backups taken should NOT be removed.
```
[the remark this answers is not in the document]
Comment (chriscchien) [reply e2e/tests/test_cases/uninstallation_checks.robot:41]: Updated, but does step 6 can automated? because test case mentioned the DR volume is in another cluster.

File: e2e/libs/uninstallation/uninstallation.py (hunk 1 above)
Comment (chriscchien) [reply e2e/libs/uninstallation/uninstallation.py:11]: Updated, we already have `e2e/libs/longhorn.py` so I add below in `e2e/libs/longhorn_deploy`
.
├── base.py
├── __init__.py
├── longhorn_deploy.py
├── longhorn_helm_chart.py
└── longhorn_kubectl.py

File: e2e/tests/test_cases/uninstallation_checks.robot (hunk 3 above)
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/tests/test_cases/uninstallation_checks.robot:41]: Yes. Since you uninstall and reinstall Longhorn, the reinstalled Longhorn is another cluster.

File: e2e/tests/test_cases/uninstallation_checks.robot [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Test Setup    Set test environment
+Test Teardown    Cleanup test resources
+
+*** Variables ***
+${LOOP_COUNT}    1
+${RETRY_COUNT}    300
+${RETRY_INTERVAL}    1
+${DATA_ENGINE}    v1
+${LONGHORN_BRANCH}    master
```
Comment (yangchiu) [inline e2e/tests/test_cases/uninstallation_checks.robot:26]: It would be ideal if the installation parameters, such as `LONGHORN_REPO_BRANCH`, `CUSTOM_LONGHORN_MANAGER_IMAGE`, `CUSTOM_LONGHORN_ENGINE_IMAGE`, could be retrieved from Jenkins environment variables, and passed to a shell script for installing Longhorn using different method, similar to what our `test_upgrade` pytest and `upgrade_longhorn.sh` do. Do you think it's possible for you to implement this?

File: e2e/libs/uninstallation/uninstallation.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+from utility.utility import logging
```
Comment (yangchiu) [inline e2e/libs/uninstallation/uninstallation.py:1]: Is this file still needed?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/keywords/longhorn.resource (hunk 2 above)
Comment (chriscchien) [reply e2e/keywords/longhorn.resource]: In `libs/longhorn_deploy/lognhorn.py` added below part because Jenkins did not have `LONGHORN_INSTALL_METHOD` on e2e pipeline, if run on local and using helm, we need to set set `LONGHORN_INSTALL_METHOD` to `helm-chart` first
```
_method = os.getenv("LONGHORN_INSTALL_METHOD", "manifest-file")
```

File: e2e/tests/test_cases/uninstallation_checks.robot (hunk 4 above)
Comment (chriscchien) [reply e2e/tests/test_cases/uninstallation_checks.robot]: Updated, added `e2e/utilities/longhorn-setup.sh` to get  environment variable to customize test images. I tested on local and Jenkins

[report.zip](https://github.com/user-attachments/files/17009715/report.zip)

image

**Comment by chriscchien** [comment]
@longhorn/qa, this ticket is ready for review.

File: e2e/tests/test_cases/uninstallation_checks.robot (hunk 4 above)
Comment (yangchiu) [reply e2e/tests/test_cases/uninstallation_checks.robot]: Why is adding `e2e/utilities/longhorn-setup.sh` necessary to obtain the environment variables for custom images? Aren't those variables already passed to the container in `pipeline/e2e/Jenkinsfile` at https://github.com/longhorn/longhorn-tests/blob/master/pipelines/e2e/Jenkinsfile#L95?

File: e2e/utilities/longhorn-setup.sh
Comment (yangchiu) [inline e2e/utilities/longhorn-setup.sh:1]: Why is adding `e2e/utilities/longhorn-setup.sh` necessary to obtain the environment variables for custom images? Aren't those variables already passed to the container in `pipeline/e2e/Jenkinsfile` at https://github.com/longhorn/longhorn-tests/blob/master/pipelines/e2e/Jenkinsfile#L95?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/utilities/longhorn-setup.sh
Comment (chriscchien) [reply e2e/utilities/longhorn-setup.sh:1]: This test case will uninstall LH and then reinstall again. The reinstall Longhorn steps in `e2e/utilities/longhorn-setup.sh` is executed in robot framework which is running in `longhorn-test` pod  ([run_longhorn_e2e_test.sh](https://github.com/longhorn/longhorn-tests/blob/master/pipelines/utilities/run_longhorn_e2e_test.sh)), thus I need pass the environments when create the test pod, otherwise the reinstall will fail because information for custom images are `None`.
If this doesn't fully answer the question, could you elaborate more on your concern? thanks.

Comment (chriscchien) [reply e2e/utilities/longhorn-setup.sh:1]: In addition, by default implementation, some Jenkins env variable did pass into instacne :
```bash
ip-10-0-1-121:/home/ec2-user # printenv
LS_COLORS=no=00:fi=00:di=01;34:ln=00;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=41;33;01:ex=00;32:*.cmd=00;32:*.exe=01;32:*.com=01;32:*.bat=01;32:*.btm=01;32:*.dll=01;32:*.tar=00;31:*.tbz=00;31:*.tgz=00;31:*.rpm=00;31:*.deb=00;31:*.arj=00;31:*.taz=00;31:*.lzh=00;31:*.lzma=00;31:*.zip=00;31:*.zoo=00;31:*.z=00;31:*.Z=00;31:*.gz=00;31:*.bz2=00;31:*.tb2=00;31:*.tz2=00;31:*.tbz2=00;31:*.xz=00;31:*.avi=01;35:*.bmp=01;35:*.dl=01;35:*.fli=01;35:*.gif=01;35:*.gl=01;35:*.jpg=01;35:*.jpeg=01;35:*.mkv=01;35:*.mng=01;35:*.mov=01;35:*.mp4=01;35:*.mpg=01;35:*.pcx=01;35:*.pbm=01;35:*.pgm=01;35:*.png=01;35:*.ppm=01;35:*.svg=01;35:*.tga=01;35:*.tif=01;35:*.webm=01;35:*.webp=01;35:*.wmv=01;35:*.xbm=01;35:*.xcf=01;35:*.xpm=01;35:*.aiff=00;32:*.ape=00;32:*.au=00;32:*.flac=00;32:*.m4a=00;32:*.mid=00;32:*.mp3=00;32:*.mpc=00;32:*.ogg=00;32:*.voc=00;32:*.wav=00;32:*.wma=00;32:*.wv=00;32:
HOSTTYPE=x86_64
LESSCLOSE=lessclose.sh %s %s
XKEYSYMDB=/usr/X11R6/lib/X11/XKeysymDB
LANG=POSIX
WINDOWMANAGER=xterm
LESS=-M -I -R
SUDO_GID=100
HOSTNAME=ip-10-0-1-121
CSHEDIT=emacs
GPG_TTY=/dev/pts/2
LESS_ADVANCED_PREPROCESSOR=no
COLORTERM=1
SUDO_COMMAND=/usr/bin/su
MACHTYPE=x86_64-suse-linux
MINICOM=-c on
NCURSES_NO_UTF8_ACS=1
OSTYPE=linux
USER=root
PAGER=less
MORE=-sl
PWD=/home/ec2-user
HOME=/root
LC_CTYPE=C.UTF-8
HOST=ip-10-0-1-121
SUDO_USER=ec2-user
XNLSPATH=/usr/X11R6/lib/X11/nls
XDG_DATA_DIRS=/usr/share
PROFILEREAD=true
SUDO_UID=1000
FROM_HEADER=
MAIL=/var/mail/root
LESSKEY=/etc/lesskey.bin
SHELL=/bin/bash
TERM=xterm-256color
LS_OPTIONS=-A -N --color=tty -T 0
SHLVL=1
MANPATH=/usr/share/man:/usr/local/man
LOGNAME=root
XDG_CONFIG_DIRS=/etc/xdg
PATH=/sbin:/usr/sbin:/usr/local/sbin:/root/bin:/usr/local/bin:/usr/bin:/bin:/usr/lib/mit/bin
G_BROKEN_FILENAMES=1
HISTSIZE=1000
CPU=x86_64
LESSOPEN=lessopen.sh %s
_=/usr/bin/printenv
```

Comment (yangchiu) [reply e2e/utilities/longhorn-setup.sh:1]: Is it possible to reuse functions from `e2e/utilities/` instead of copying and pasting the entire `longhorn-setup.sh` file?

Also, can we pass custom component image versions as parameters (e.g., `-v CUSTOM_LONGHORN_INSTANCE_MANAGER_IMAGE:longhornio/longhorn-instance-manager:master-head`) similar to what we did [in `test_upgrade` pytest.](https://github.com/longhorn/longhorn-tests/blob/master/test_framework/scripts/longhorn-setup.sh#L378) so we don't have to modify the `test.yaml` file?

**Comment by chriscchien** [comment]
@longhorn/qa , this ticket updated and ready for revirew, thank you.

File: e2e/libs/k8s/k8s.py [hunk 6]
```diff
@@ -71,3 +72,74 @@ def wait_all_pods_evicted(node_name):
...
+
+def wait_namespace_terminated(namespace):
+    retry_count, retry_interval = get_retry_count_and_interval()
+    api = client.CoreV1Api()
+    for i in range(retry_count):
+        try:
+            target_namespace = api.read_namespace(name=namespace)
+            target_namespace_status = target_namespace.status.phase
+            logging(f"Waiting for namespace {target_namespace.metadata.name} terminated, current status is {target_namespace_status} retry ({i}) ...")
+        except ApiException:
```
Comment (yangchiu) [inline e2e/libs/k8s/k8s.py:108]: Should it also verify if `e.status == 404`?

File: pipelines/utilities/run_longhorn_e2e_test.sh [hunk 7]
```diff
@@ -31,6 +32,30 @@ run_longhorn_e2e_test(){
...
+  yq e -i 'select(.spec.containers[0].env != null).spec.containers[0].env += {"name": "CUSTOM_LONGHORN_ENGINE_IMAGE", "value": "'${CUSTOM_LONGHORN_ENGINE_IMAGE}'"}' "${CUSTOM_LONGHORN_ENGINE_IMAGE}"
+  yq e -i 'select(.spec.containers[0].env != null).spec.containers[0].env += {"name": "CUSTOM_LONGHORN_INSTANCE_MANAGER_IMAGE", "value": "'${CUSTOM_LONGHORN_INSTANCE_MANAGER_IMAGE}'"}' "${LONGHORN_TESTS_MANIFEST_FILE_PATH}"
+  yq e -i 'select(.spec.containers[0].env != null).spec.containers[0].env += {"name": "CUSTOM_LONGHORN_SHARE_MANAGER_IMAGE", "value": "'${CUSTOM_LONGHORN_SHARE_MANAGER_IMAGE}'"}' "${LONGHORN_TESTS_MANIFEST_FILE_PATH}"
+  yq e -i 'select(.spec.containers[0].env != null).spec.containers[0].env += {"name": "CUSTOM_LONGHORN_BACKING_IMAGE_MANAGER_IMAGE", "value": "'${CUSTOM_LONGHORN_BACKING_IMAGE_MANAGER_IMAGE}'"}' "${LONGHORN_TESTS_MANIFEST_FILE_PATH}"
+  yq e -i 'select(.spec.containers[0].env != null).spec.containers[0].env += {"name": "LONGHORN_INSTALL_METHOD", "value": "'${LONGHORN_INSTALL_METHOD}'"}' "${LONGHORN_TESTS_MANIFEST_FILE_PATH}"
+
+  set +x
+  if [[ "${LONGHORN_TEST_CLOUDPROVIDER}" == "aws" ]]; then
```
Comment (yangchiu) [inline pipelines/utilities/run_longhorn_e2e_test.sh:47]: This part should be removed as it's based on outdated content.

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/utilities/longhorn-install.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+set -x
+
+source ../test_framework/scripts/longhorn-setup.sh
```
[the remark this answers is not in the document]
Comment (chriscchien) [reply e2e/utilities/longhorn-install.sh]: updated

```
source ../pipelines/utilities/longhorn_ui.sh
source ../pipelines/utilities/create_longhorn_namespace.sh
source ../pipelines/utilities/install_backupstores.sh
source ../pipelines/utilities/longhorn_status.sh
source ../pipelines/utilities/longhorn_helm_chart.sh
source ../pipelines/utilities/create_aws_secret.sh
source ../pipelines/utilities/longhorn_manifest.sh
```

File: e2e/libs/k8s/k8s.py (hunk 6 above)
Comment (chriscchien) [reply e2e/libs/k8s/k8s.py]: updated

```python
try:
            target_namespace = api.read_namespace(name=namespace)
            target_namespace_status = target_namespace.status.phase
            logging(f"Waiting for namespace {target_namespace.metadata.name} terminated, current status is {target_namespace_status} retry ({i}) ...")
        except ApiException as e:
            if e.status == 404:
                logging(f"Namespace {namespace} successfully terminated.")
                return
            else:
                logging(f"Error while fetching namespace {namespace} status: {e}")
```

File: pipelines/utilities/run_longhorn_e2e_test.sh (hunk 7 above)
Comment (chriscchien) [reply pipelines/utilities/run_longhorn_e2e_test.sh]: Updated

File: e2e/utilities/longhorn-install.sh [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+source ../pipelines/utilities/longhorn_manifest.sh
+
+# create and clean tmpdir
+TMPDIR="/tmp/longhorn"
+mkdir -p ${TMPDIR}
+rm -rf "${TMPDIR}/"
+
+LONGHORN_NAMESPACE="longhorn-system"
+
+install_longhorn_by_chart(){
```
Comment (yangchiu) [inline e2e/utilities/longhorn-install.sh:20]: Can we reuse functions in `longhorn_helm_chart.sh` here?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/utilities/longhorn-install.sh (hunk 9 above)
Comment (chriscchien) [reply e2e/utilities/longhorn-install.sh:19]: @yangchiu, Updated!

**Review by yangchiu — APPROVED** [review]
LGTM
